src/nnbench/reporter/file.py

```diff
@@ -0,0 +1,205 @@
+
+    @classmethod
+    def register(cls, file_type: str) -> None:
+        """Registers a parser for a specific file type.
+
+        Args:
+            `file_type:` The file type (string)
+        """
+        parsers[file_type] = cls
+
+    @staticmethod
+    def get_parser(file_type: str):
+        """Gets the registered parser for a file type.
+
+        Args:
+            `file_type:` The file type (string)
+
+        Returns:
+        --------
+            The registered RecordParser, or None if not found.
+        """
+        return parsers.get(file_type)
+
+
+class JsonParser(Parser):
+    def parse_file(self, records: str) -> List[dict]:
+        import json
+
+        try:
+            return json.loads(records if records else "[]")
+        except json.JSONDecodeError:
+            raise ValueError("Unexpected records passed")
+
+    def write_records(
+        self, parsed_records: Any[BenchmarkRecord] | None, record: BenchmarkRecord
+    ) -> str:
+        import json
+
+        parsed_records.append(record)
+        return json.dumps(parsed_records)
+
+
+class YamlParser(Parser):
+    def parse_file(self, records: str) -> List[dict]:
+        import yaml
+
+        return yaml.safe_load(records) if records else []
+
+    def write_records(
+        self, parsed_records: Any[BenchmarkRecord] | None, record: BenchmarkRecord
+    ) -> str:
+        import yaml
+
+        parsed_records.append(record)
+        for element in record["benchmarks"]:
+            element["value"] = float(element["value"])
+        return yaml.dump(parsed_records)
+
+
+# Register custom parsers here
+parsers = {"json": JsonParser, "yaml": YamlParser}
+
+
+def parse_records(records: str, file_type: str) -> Any:
+    """Parses records based on the specified file type.
+
+    This function retrieves and calls the registered parser for
+    the given file type.
+
+    Args:
+        `records:` A list or iterator of record strings.
+        `file_type:` The file type (string).
+
+    Returns:
+        A list of parsed records.
+    """
+
+    parser = Parser.get_parser(file_type)
+    if parser is None:
+        raise ValueError(f"Unsupported file type: {file_type}")
+
+    return parser().parse_file(records)
+
+
+def append_record_to_records(parsed_records: Any, record: BenchmarkRecord, file_type: str) -> str:
+    """Appends a record to the list based on the file type.
+
+    This function first parses the record using the appropriate parser
+    and then appends it to the `parsed_records`.
+
+    Args:
+        `records:` A list of parsed records.
+        `record:` The record to append.
+        `file_type:` The file type (string).
+    """
+
+    parser = Parser.get_parser(file_type)
+    if parser is None:
+        raise ValueError(f"Unsupported file type: {file_type}")
+
+    return parser().write_records(parsed_records, record)
+
+
+class FileReporter(BenchmarkReporter):
+    def __init__(self, dir: str):
+        self.dir = dir
```

Comment:
`dir` is a builtin name, so I would change this to e.g. `directory` or `result_dir`.